src/identity/transaction_manager.rs

```diff
@@ -0,0 +1,545 @@
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::Duration;
+
+use anyhow::Context;
+use async_trait::async_trait;
+use chrono::Utc;
+use ethers::abi::RawLog;
+use ethers::addressbook::Address;
+use ethers::contract::EthEvent;
+use ethers::middleware::Middleware;
+use ethers::prelude::{Log, Topic, ValueOrArray, U256};
+use tracing::{error, info, instrument};
+
+use crate::config::Config;
+use crate::contracts::abi::{BridgedWorldId, RootAddedFilter, TreeChangeKind, TreeChangedFilter};
+use crate::contracts::scanner::BlockScanner;
+use crate::contracts::IdentityManager;
+use crate::database::query::DatabaseQuery;
+use crate::database::types::{BatchEntry, BatchType};
+use crate::database::Database;
+use crate::ethereum::{Ethereum, ReadProvider};
+use crate::identity_tree::{Canonical, Intermediate, TreeVersion, TreeWithNextVersion};
+use crate::prover::Prover;
+use crate::utils::index_packing::pack_indices;
+use crate::utils::retry_tx;
+
+pub type TransactionId = String;
+
+impl From<TransactionId> for crate::ethereum::write::TransactionId {
```

Comment:
`TransactionId` is just a tiny wrapper around `String`, I don't think there's a need for this alias here

Reply:
It was just to not depend on external library with API here. In offchain mode it would be strange to return ethereum::write::TransactionId instead of transaction id from identity module. But I will left decision for others here. ;)

Reply:
But it's not an external library we define `TransactionId` in this project

Reply:
I left single TransactionId - the one that I added. I removed the one from ethereum crate just to not depend on "ethereum" for off chain mode as it would be odd to me. :)

Reply:
Sounds good to me